File: src/main/java/com/foryourlife/shared/email/DispatchNotification.java
package com.foryourlife.shared.email;

import com.sendgrid.Method;
import com.sendgrid.Request;
import com.sendgrid.Response;
import com.sendgrid.SendGrid;
import com.sendgrid.helpers.mail.Mail;
import com.sendgrid.helpers.mail.objects.Content;
import com.sendgrid.helpers.mail.objects.Email;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.IOException;

@Service
public class DispatchNotification {
    private final SendGrid sendGrid;
    private final String fromEmail;

    public DispatchNotification(
            @Value("${com.email.sendgrid.api-key}") String sendgridApiKey,
            @Value("${com.email.sendgrid.from-email}") String fromEmail) {
        this.sendGrid = new SendGrid(sendgridApiKey);
        this.fromEmail = fromEmail;
    }

    public void send(String toEmail, String subject, String message) {
        try {
            Email to = new Email(toEmail);
            Email from = new Email(fromEmail);

            Content content = new Content("text/html", message);
            Mail mail = new Mail(from, subject, to, content);

            Request request = new Request();
            request.setMethod(Method.POST);
            request.setEndpoint("mail/send");
            request.setBody(mail.build());

            Response response = sendGrid.api(request);

            System.out.println("Correo enviado. Código de estado: " + response.getStatusCode());
            System.out.println("Respuesta de SendGrid: " + response.getBody());

        } catch (IOException e) {
            System.out.println("Error al enviar el correo: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: src/main/java/com/foryourlife/shared/email/DispatchNotification.kt
package com.foryourlife.shared.email

import com.sendgrid.Method
import com.sendgrid.Request
import com.sendgrid.SendGrid
import com.sendgrid.helpers.mail.Mail
import com.sendgrid.helpers.mail.objects.Attachments
import com.sendgrid.helpers.mail.objects.Content
import com.sendgrid.helpers.mail.objects.Email
import org.springframework.stereotype.Service
import org.thymeleaf.TemplateEngine
import org.thymeleaf.context.Context
import java.io.*
import java.nio.charset.StandardCharsets
import java.time.LocalDate
import java.time.format.TextStyle
import java.util.Base64
import java.util.Locale


@Service
class DispatchNotification(
    val email: Email,
    val sendGrid: SendGrid,
    private val templateEngine: TemplateEngine,
) {
    fun send(toEmail: String, subject: String, message: String) {
        try {


            val now = LocalDate.now()
            val day = now.dayOfMonth
            val month = now.month.getDisplayName(
                TextStyle.FULL, Locale("es")
            )
            val toEmail = Email(toEmail)
            val context = Context().apply {
                setVariables(
                    mapOf(
                        "message" to message,
                    )
                )
            }

            val htmlContent = templateEngine.process(
                "email-template.html", context
            )
            val content = Content(
                "text/html", htmlContent
            )
            val mail = Mail(
                email, subject, toEmail, content
            )

            val request = Request()
            request.method = Method.POST
            request.endpoint = "mail/send"
            request.body = mail.build()

            val response = sendGrid.api(request)

            println("Correo enviado. Código de estado: ${response.statusCode}")
            println("Respuesta de SendGrid: ${response.body}")

        } catch (e: Exception) {
            println("Error al enviar el correo: ${e.message}")
            e.printStackTrace()
        }
    }

    private fun createAttachment(file: File): Attachments {
        val encodedFileContent: ByteArray = Base64.getEncoder().encode(file.readBytes())
        val attachment = Attachments()
        attachment.disposition = "attachment"
        attachment.type = file.extension
        attachment.filename = file.name
        attachment.content = String(
            encodedFileContent,
            StandardCharsets.UTF_8
        )
        return attachment
    }
}
